### PR TITLE
cleanup parseNumParam

### LINKED ISSUE #1469: cleanup parseNumParam (CommandHandler)
function signature is bad - we should make it throw now that all handlers are dealing with exceptions properly.
New signature should probably return an `optional<>` and takes as parameter a string in case of error

### PR DESCRIPTION
**Pull request by marta-lokhova**
resolves #1469 

* Update the signature to return optional<T> and removed redundant arguments
* Throw an exception in case arg can't be loaded; Note that if a number param followed by garbage (e.g. non-numeric characters), an exception will be thrown also
* tested affected endpoints with various parameters my running stellar-core locally

### CODE REVIEW

File: src/main/CommandHandler.cpp [hunk 1]
```diff
@@ -364,36 +364,27 @@ enum class Requirement
 };
 
 template <typename T>
-bool
+optional<T>
 parseNumParam(std::map<std::string, std::string> const& map,
-              std::string const& key, T& val, std::string& retStr,
-              Requirement requirement, bool& valueUpdated)
+              std::string const& key, T& val)
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: you don't need to pass in `val` anymore.
If people want to do default values, I would expect callers to do things like
```
int n = 1000; // default value
auto v = parseNumParam<int>(m, "n");
if (v)
{
    n = *v;
}
```
non optional case is simply:
```
int n = *parseNumParam<int>(m, "n");
```

File: src/main/CommandHandler.cpp [hunk 2]
```diff
@@ -364,36 +364,27 @@ enum class Requirement
...
-    valueUpdated = false;
     auto i = map.find(key);
     if (i != map.end())
     {
         std::stringstream str(i->second);
         str >> val;
-        if (val == 0)
+
+        // Throw an error if not all bytes were loaded into `val`
+        if (str.fail() || str.rdbuf()->in_avail() > 0)
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: isn't the second condition just `!str.eof()`?

File: src/main/CommandHandler.cpp [hunk 3]
```diff
@@ -725,17 +710,8 @@ CommandHandler::upgrades(std::string const& params, std::string& retStr)
...
-            {
-                retStr = (fmt::MemoryWriter()
-                          << fmt::format("could not parse {}: '{}'\n", name,
-                                         retMap[name])
-                          << retStr)
-                             .str();
-            }
-            else if (updated)
+            auto res = parseNumParam(retMap, name, v);
+            if (res)
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: isn't that code equivalent to `f = parseNumParam(...)` ?
If so, this lambda is now obsolete

File: src/main/CommandHandler.cpp [hunk 4]
```diff
@@ -950,8 +926,7 @@ CommandHandler::setcursor(std::string const& params, std::string& retStr)
 
     uint32 cursor;
 
-    if (!parseNumParam(map, "cursor", cursor, retStr,
-                       Requirement::REQUIRED_REQ))
+    if (!parseNumParam(map, "cursor", cursor))
     {
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: this code can be deleted

Comment (marta-lokhova) [reply src/main/CommandHandler.cpp]: Don't we need this check to gracefully handle the case when `cursor` is not passed in the request? With the updated logic based on your comments, it would be something like 
```
    uint32 cursor;
    auto res = parseNumParam(map, "cursor");

    if (!res)
    {
        retStr = "Invalid cursor";
        return;
    }
    else
    {
        cursor = *res;
    }
```

Comment (MonsieurNicolas) [reply src/main/CommandHandler.cpp]: If it’s a required parameter this should be taken care of by `parseNumParam`

File: src/main/CommandHandler.cpp (hunk 3 above)
Comment (marta-lokhova) [reply src/main/CommandHandler.cpp]: I think you're right. `reset()` would result in the same behavior as assigning `nullopt` (which is `shared_ptr`() under the hood)

**Comment by marta-lokhova** [comment]
Thanks for reviewing! I've updated the PR based on the comments.

File: src/main/CommandHandler.cpp [hunk 5]
```diff
@@ -363,28 +363,53 @@ enum class Requirement
     REQUIRED_REQ
 };
 
-template <typename T>
-optional<T>
+optional<uint32>
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: why did you remove the template?

File: src/main/CommandHandler.cpp [hunk 6]
```diff
@@ -363,28 +363,53 @@ enum class Requirement
...
-        return make_optional<T>(val);
+        return make_optional<uint32>(val);
     }
 
-    return nullopt<T>();
+    return nullopt<uint32>();
+}
+
+optional<uint32>
+parseNumParam(std::map<std::string, std::string> const& map,
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: After looking at the changes overall, I don't think this contract makes sense to me:
we already have a `parseNumParam` that returns an `optional<T>` (that can be used for anything).

From the use cases point of view, I think it would probably make more sense to rename the optional version to `maybeParseNumParam` and add a dedicated method to deal with the non-optional case:
in this case, the method's signature would not return an optional, and there is no default value.

Then it probably makes sense to update the `maybeParseNumParam` that returns an optional take care of the default value (and have the other one call it with a default value internally).

File: src/main/CommandHandler.cpp [hunk 7]
```diff
@@ -363,28 +363,53 @@ enum class Requirement
...
+              std::string const& key, uint32& defaultVal,
+              Requirement requirement)
+{
+    auto val = parseNumParam(map, key);
+    if (val)
+    {
+        if (requirement == Requirement::OPTIONAL_REQ)
+        {
+            defaultVal = *val;
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: that logic doesn't make sense to me: it's updating `defaultVal`?

**Comment by MonsieurNicolas** [comment]
cool, can you rebase and squash commit and we'll get this in

**Comment by marta-lokhova** [comment]
Sure, it's been updated.

**Comment by MonsieurNicolas** [comment]
r+ 4a9ead9d06f98e188742aaae9e58b8675bfb354c

**Comment by MonsieurNicolas** [comment]
@latobarita: retry

**Comment by MonsieurNicolas** [comment]
hit by #1488

**Comment by MonsieurNicolas** [comment]
@latobarita: retry

**Comment by MonsieurNicolas** [comment]
maybe running that test with a combination of the shell commands `taskset ...` and `cpulimit -i ...`  can help diagnose this outside of travis? @vogel
